### PR TITLE
Run notebooks on RTD with sphinx-gallery and add WorkGraph tutorials

### PR DESCRIPTION
**Pull request by agoscinski**
Similar as in aiida-workgraph https://github.com/aiidateam/aiida-workgraph/pull/234 I would like to integrate an execution of the notebooks as the notebooks become quickly outdated otherwise. This simplifies the update process of the notebooks as they will be automatically run with each CI, so errors can be immediately seen and then fixed. Since we already adapted the tutorials from the aiida-workgraph to the sphinx-scripts and I would like to move these tutorials out of aiida-workgraph to this repo as fast as possible I focused on moving these as a first example. Nevertheless I would like to adapt all tutorials to be executable in the future.

I tried myst_nb for the execution as it was already provided by the repo, but I ran into issues that ran explained below, so I used the same strategy as in aiida-workgraph by using sphinx-gallery. 

Note that the sections `Writing workflows` and `Writing workflows with WorkGraph` (the new section including the WorkGraph) have overlaps that could be better resolved, but I did not want to spend the time in the first PR setting up the project structure, since I feel like this could become very time consuming.

Also Note that the CI link checker has been merged to RTD. Since for link checking one also needs to run the tutorials, I thought this is less maintenance as one just needs to update the RTD setup. This also does not affect the building time, because on the second build when creating the html files, the files from the first build can be Reused. Compare the times of https://readthedocs.org/projects/aiida-tutorials/builds/25605020/ of two builds (linkcheck plus html) and https://readthedocs.org/projects/aiida-tutorials/builds/25603891/ one build (only html)

### CODE REVIEW

**Comment by agoscinski** [comment]
I tried using myst_nb but I get `WARNING: No source code lexer found for notebook cell 1 [myst-nb.lexer]` warnings which I cannot find any solution to online. It seems that I need to change markdown part in the cells, but I am really not sure. Also when an error is thrown it is written into a file and not printed out, which is really crucial for debugging runs on RTD. So I am switching to sphinx-gallery.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,6 +1,11 @@
 name: continuous-integration
 
-on: [push, pull_request]
+on:
+  push:
+    # only pushes to main trigger
+    branches: [main]
+  pull_request:
+    # always triggered
```
Comment (agoscinski) [inline .github/workflows/ci.yml:8]: This change is unrelated to this PR. It stops that the CI is run two time for each PR. I will put this in a separate PR

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -11,29 +16,8 @@ jobs:
 
     steps:
     - uses: actions/checkout@v2
-    - name: Set up Python 3.12
+    - name: Set up Python 3.11
```
Comment (agoscinski) [inline .github/workflows/ci.yml:19]: To build aiida-core in CI, I needed to use Python 3.11. For pre-commit it is not needed but I think we should just stick with Python 3.11 if there are issues with Python 3.12.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -11,29 +16,8 @@ jobs:
...
       uses: actions/setup-python@v2
       with:
-        python-version: 3.12
+        python-version: 3.11
     - uses: pre-commit/action@v2.0.0
-
-  build:
```
Comment (agoscinski) [inline .github/workflows/ci.yml:20]: This is now run in the RTD, check out the `.readthedocs.yml`

File: .readthedocs.yml [hunk 4]
```diff
@@ -19,7 +33,10 @@ sphinx:
...
+conda:
+  environment: environment.yml
+
 # Optionally set the version of Python and requirements required to build your docs
-python:
-  install:
-    - requirements: requirements.txt
+#python:
+#  install:
+#    - requirements: requirements.txt
```
Comment (agoscinski) [inline .readthedocs.yml]: This is not needed anymore since it is in the `environment.yml`
```suggestion
```

File: README.md [hunk 5]
```diff
@@ -26,17 +26,19 @@ If you have a question, feel free to just [open an issue](https://github.com/aii
...
 git clone https://github.com/aiidateam/aiida-tutorials.git
 cd aiida-tutorials
-pip install -r requirements.txt
+conda env create --quiet --name aiida-tutorials --file environment.yml
+conda activate aiida-tutorials
 pre-commit install   # enable pre-commit hooks (optional)
-cd docs/
-make
+make -C docs html # to build docs
+make -C docs html linkcheck # to run link checks
```
Comment (agoscinski) [inline README.md]:
```suggestion
make -C docs linkcheck # to run link checks (only for dev)
```

File: docs/conf.py [hunk 6]
```diff
@@ -68,7 +69,7 @@
 ipython_mplbackend = ""
 
 copybutton_selector = "div:not(.no-copy)>div.highlight pre"
-copybutton_prompt_text = ">>> |\.\.\. |\$ |In \[\d*\]: | {2,5}\.\.\.: | {5,8}: "
+copybutton_prompt_text = ">>> |... |$ |In [d*]: | {2,5}...: | {5,8}: "
```
Comment (agoscinski) [inline docs/conf.py:72]: The escapes raised some warnings

File: docs/conf.py [hunk 7]
```diff
@@ -391,15 +392,14 @@
...
-
-def setup(app):
-    """Setup function called by sphinx."""
-    app.add_css_file("css/custom.css")
-
+linkcheck_ignore = [
+    "https://www.big-map.eu/",
+    r".*concept/index.html",
+    r".*howto/index.html",
+    "http://127.0.0.1:8000/workgraph",
```
Comment (agoscinski) [inline docs/conf.py:399]: Here we ignore workgraph related files

File: docs/conf.py [hunk 8]
```diff
@@ -391,15 +392,14 @@
...
+]
 
-# We are not installing a full aiida environment
+# we don't want to run the notebook
```
Comment (agoscinski) [inline docs/conf.py]: changing back, remains from my try to use myst_nb
```suggestion
# We are not installing a full aiida environment
```

File: docs/conf.py [hunk 9]
```diff
@@ -416,3 +439,57 @@ def setup(app):
         universal_newlines=True,
     )
 )
+
+import shutil
+from pathlib import Path
+
+
+def copy_html_files(app, exception):
```
Comment (agoscinski) [inline docs/conf.py:447]: copy paste from aiida-workgraph to move html files from the sphinx gallery folder to the sphinx source, so we can display the widgets

File: docs/gallery/sections/writing_workflows_with_workgraph/autogen/GALLERY_HEADER.rst [hunk 10]
```diff
@@ -0,0 +1,5 @@
+:orphan:
+
+================================
+Running workflows with WorkGraph
+================================
```
Comment (agoscinski) [inline docs/gallery/sections/writing_workflows_with_workgraph/autogen/GALLERY_HEADER.rst:5]: sphinx-gallery requires a `GALLERY_HEADER.rst` but we do not use if anywhere therefore the `:orphan:` at the top to surppres warnings

File: docs/gallery/sections/writing_workflows_with_workgraph/autogen/eos.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
+"""
```
Comment (agoscinski) [inline docs/gallery/sections/writing_workflows_with_workgraph/autogen/eos.py:1]: copy paste from aiida-workgraph tutorials as all sphinx-gallery scripts

File: requirements.txt [hunk 12]
```diff
@@ -4,4 +4,9 @@ sphinx-book-theme~=1.1.3
 sphinx-copybutton~=0.5.2
 sphinx-panels~=0.4.1
 sphinx-tabs~=3.4.5
+sphinx-gallery
```
Comment (agoscinski) [inline requirements.txt]:
```suggestion
sphinx-gallery~=0.17.1
```

File: requirements.txt [hunk 13]
```diff
@@ -4,4 +4,9 @@ sphinx-book-theme~=1.1.3
...
 myst-nb~=1.1.1
+# to run notebooks
+aiida-quantumespresso
+aiida-pseudo
+aiida-workgraph[widget]
```
Comment (agoscinski) [inline requirements.txt:12]: I think aiida-workgraph we do not want to pin, but I am not sure about the others

File: docs/sections/writing_workflows_with_workgraph/index.rst [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    :column: col-lg-12
+
+    .. link-button:: autogen/zero_to_hero
+        :type: ref
+        :text: Zero to hero
+        :classes: btn-light text-left stretched-link font-weight-bold
+    ^^^^^^^^^^^^
+
+    A short module on how to write the basic type of workflows in AiiDA: work functions.
+    The module also revises the usage of calculation functions to add simple Python functions to the provenance.
```
Comment (agoscinski) [inline docs/sections/writing_workflows_with_workgraph/index.rst:27]: This text needs some rework, maybe one of the reviewer has some ideas

File: docs/sections/writing_workflows_with_workgraph/index.rst [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    :column: col-lg-12
+
+    .. link-button:: autogen/qe
+        :type: ref
+        :text: Computational materials science
+        :classes: btn-light text-left stretched-link font-weight-bold
+    ^^^^^^^^^^^^
+
+    A step-by-step introduction to the basics of writing work chains in AiiDA.
+    After completing this module, you will be ready to start writing your own scientific workflows!
```
Comment (agoscinski) [inline docs/sections/writing_workflows_with_workgraph/index.rst:53]: This text needs some rework, maybe one of the reviewer has some ideas

**Comment by superstar54** [comment]
Hi @agoscinski , thanks for the work.

> I tried myst_nb for the execution as it was already provided by the repo, but I ran into issues that ran explained below, so I used the same strategy as in aiida-workgraph by using sphinx-gallery.

What's the issue? Can we fix it? It's better to use the same method for all, to make it easy to maintain.

### Refactor the notebooks
There are duplicate contents in the three notebooks:
  -  EOS workgraph in both zero-to-hero and the eos notebooks.
  - atomization energy workgraph in both zero-to-hero and computational materials science notebooks.
I would suggest replacing the `zero-to-hero` with workgraph basic.

### Compare with WorkChain
Since the previous section is WorkChain, and it also shows examples of simple add and real-world EOS:
- There are some calcfunctions already defined in the WorkChain section; use them, if possible, to ensure consistency.
- Mention the WorkChain version, or try to compare them with WorkGraph, so that users know why WorkGraph makes it simple.


#### Use same style as the other section
The Workchain section is well-written, explaining steps step by step and providing extensive descriptions. It also has `Note`, `Important` and `explanation`. I suggest using the same style to make the website more uniform.
Also, the font size of the title/subtitle, etc., should be the same.

To summarize, I would suggest rewriting the WorkGraph section to fit into the aiida-tutorial better.

File: docs/gallery/sections/writing_workflows_with_workgraph/autogen/zero_to_hero.py [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+#
+# What's Next
+# ===========
+#
+# +-----------------------------------------------------+-----------------------------------------------------------------------------+
+# | `Concepts <../../concept/index.html>`_              | A brief introduction of WorkGraph’s main concepts.                          |
+# +-----------------------------------------------------+-----------------------------------------------------------------------------+
+# | `HowTo <../../howto/index.html>`_                   | Advanced topics, e.g., flow control using `if`, `while`, and `context`.     |
+# +-----------------------------------------------------+-----------------------------------------------------------------------------+
+#
```
Comment (superstar54) [inline docs/gallery/sections/writing_workflows_with_workgraph/autogen/zero_to_hero.py:527]: This is not valid anymore; maybe remote it.
Similar to the WorkChain, we may add some exercises.

**Review by superstar54 — CHANGES_REQUESTED** [review]

**Comment by agoscinski** [comment]
> What's the issue? Can we fix it? It's better to use the same method for all, to make it easy to maintain.

Hey Xing I wrote it in the second post.

> I tried using myst_nb but I get `WARNING: No source code lexer found for notebook cell 1 [myst-nb.lexer]` warnings which I cannot find any solution to online. It seems that I need to change markdown part in the cells, but I am really not sure. Also when an error is thrown it is written into a file and not printed out, which is really crucial for debugging runs on RTD. So I am switching to sphinx-gallery.

Note that `myst_nb` is also only used for one notebook here and not well integrated.

---

Thanks for the suggestions Xing, but this PR is not intended to refactor the existing examples as I mentioned.

> Note that the sections Writing workflows and Writing workflows with WorkGraph (the new section including the WorkGraph) have overlaps that could be better resolved, but I did not want to spend the time in the first PR setting up the project structure, since I feel like this could become very time consuming.

We can put your comments in an issue.
